Compiler crash using `-coverage-out` with polymorphic extension methods

## Compiler version

3.2.0-RC2
3.2.1-RC1-bin-20220717-5c43324-NIGHTLY
fa9315923023daa3c9e524ba734bf066482cddbd from #15648

## Minimized code

```scala
extension (r: AnyRef)
  def foo[A](x: String): A = ???
  def bar[A](x: String): A = foo(x)
```

## Output (click arrow to expand)

```
scalac -coverage-out:. test.scala
```

<details>

```scala
exception while typing {
  scala.runtime.coverage.Invoker.invoked(1, ".")
  foo(r)
} of class class dotty.tools.dotc.ast.Trees$Block # -1
exception while typing {
  scala.runtime.coverage.Invoker.invoked(1, ".")
  foo(r)
}[Nothing] of class class dotty.tools.dotc.ast.Trees$TypeApply # -1
exception while typing {
  scala.runtime.coverage.Invoker.invoked(1, ".")
  foo(r)
}[Nothing](x) of class class dotty.tools.dotc.ast.Trees$Apply # -1
exception while typing {
  scala.runtime.coverage.Invoker.invoked(2, ".")
  {
    scala.runtime.coverage.Invoker.invoked(1, ".")
    foo(r)
  }[Nothing](x)
} of class class dotty.tools.dotc.ast.Trees$Block # -1
exception while typing {
  scala.runtime.coverage.Invoker.invoked(3, ".")
  {
    scala.runtime.coverage.Invoker.invoked(2, ".")
    {
      scala.runtime.coverage.Invoker.invoked(1, ".")
      foo(r)
    }[Nothing](x)
  }
} of class class dotty.tools.dotc.ast.Trees$Block # -1
exception while typing extension (r: AnyRef) def bar[A](x: String): A = 
  {
    scala.runtime.coverage.Invoker.invoked(3, ".")
    {
      scala.runtime.coverage.Invoker.invoked(2, ".")
      {
        scala.runtime.coverage.Invoker.invoked(1, ".")
        foo(r)
      }[Nothing](x)
    }
  } of class class dotty.tools.dotc.ast.Trees$DefDef # -1
exception while typing @SourceFile("test.scala") final module class test$package() extends Object() {
  private def writeReplace(): AnyRef = 
    new scala.runtime.ModuleSerializationProxy(classOf[test$package.type])
  extension (r: AnyRef) def foo[A](x: String): A = 
    {
      scala.runtime.coverage.Invoker.invoked(0, ".")
      ???
    }
  extension (r: AnyRef) def bar[A](x: String): A = 
    {
      scala.runtime.coverage.Invoker.invoked(3, ".")
      {
        scala.runtime.coverage.Invoker.invoked(2, ".")
        {
          scala.runtime.coverage.Invoker.invoked(1, ".")
          foo(r)
        }[Nothing](x)
      }
    }
} of class class dotty.tools.dotc.ast.Trees$TypeDef # -1
exception while typing package <empty> {
  final lazy module val test$package: test$package = new test$package()
  @SourceFile("test.scala") final module class test$package() extends Object() {
    private def writeReplace(): AnyRef = 
      new scala.runtime.ModuleSerializationProxy(classOf[test$package.type])
    extension (r: AnyRef) def foo[A](x: String): A = 
      {
        scala.runtime.coverage.Invoker.invoked(0, ".")
        ???
      }
    extension (r: AnyRef) def bar[A](x: String): A = 
      {
        scala.runtime.coverage.Invoker.invoked(3, ".")
        {
          scala.runtime.coverage.Invoker.invoked(2, ".")
          {
            scala.runtime.coverage.Invoker.invoked(1, ".")
            foo(r)
          }[Nothing](x)
        }
      }
  }
} of class class dotty.tools.dotc.ast.Trees$PackageDef # -1
exception occurred while compiling test.scala
java.lang.AssertionError: assertion failed while compiling test.scala
Exception in thread "main" java.lang.AssertionError: assertion failed
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:11)
	at dotty.tools.dotc.transform.Erasure$Boxing$.adaptToType(Erasure.scala:388)
	at dotty.tools.dotc.transform.Erasure$Typer.adapt(Erasure.scala:1053)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2987)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2991)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3107)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1062)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2867)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2922)
	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:126)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2987)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2991)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3107)
	at dotty.tools.dotc.transform.Erasure$Typer.typedTypeApply(Erasure.scala:807)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2878)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2922)
	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:126)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2987)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2991)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3107)
	at dotty.tools.dotc.transform.Erasure$Typer.typedApply(Erasure.scala:826)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2859)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2922)
	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:126)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2987)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2991)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3107)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1062)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2867)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2922)
	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:126)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2987)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2991)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3107)
	at dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1062)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2867)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2922)
	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:126)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2987)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2991)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3107)
	at dotty.tools.dotc.typer.Typer.$anonfun$48(Typer.scala:2295)
	at dotty.tools.dotc.typer.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:248)
	at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:2295)
	at dotty.tools.dotc.transform.Erasure$Typer.typedDefDef(Erasure.scala:943)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2835)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2921)
	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:126)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2987)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2991)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3013)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3063)
	at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1041)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2500)
	at dotty.tools.dotc.transform.Erasure$Typer.typedClassDef(Erasure.scala:1031)
	at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:2847)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2851)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2921)
	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:126)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2987)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2991)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3013)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3063)
	at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1041)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2629)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2892)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2922)
	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:126)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2987)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:2991)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3107)
	at dotty.tools.dotc.transform.Erasure.run(Erasure.scala:144)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:311)
	at scala.collection.immutable.List.map(List.scala:246)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:312)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:234)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1328)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:245)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:253)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:262)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:262)
	at dotty.tools.dotc.Run.compileSources(Run.scala:186)
	at dotty.tools.dotc.Run.compile(Run.scala:170)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:35)
	at dotty.tools.dotc.Driver.process(Driver.scala:195)
	at dotty.tools.dotc.Driver.process(Driver.scala:163)
	at dotty.tools.dotc.Driver.process(Driver.scala:175)
	at dotty.tools.dotc.Driver.main(Driver.scala:205)
	at dotty.tools.dotc.Main.main(Main.scala)
```
</details>

/cc @TheElectronWill

--- a/compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala
+++ b/compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala
@@ -101,7 +101,24 @@ class InstrumentCoverage extends MacroTransform with IdentityDenotTransformer:
 
           // (fun)[args]
           case TypeApply(fun, args) =>
-            cpy.TypeApply(tree)(transform(fun), args)
+            val tfun = transform(fun)
+            tfun match
+              case InstrumentCoverage.InstrumentedBlock(invokeCall, expr) =>
+                // expr[T] shouldn't be transformed to
+                // {invoked(...), expr}[T]
+                //
+                // but to
+                // {invoked(...), expr[T]}
+                //
+                // This is especially important for trees like (expr[T])(args),
+                // for which the wrong transformation crashes the compiler.
+                // See tests/coverage/pos/PolymorphicExtensions.scala
+                Block(
+                  invokeCall :: Nil,
+                  cpy.TypeApply(tree)(expr, args)
+                )
+              case _ =>
+                cpy.TypeApply(tree)(tfun, args)
 
           // a.b
           case Select(qual, name) =>
@@ -242,12 +259,12 @@ class InstrumentCoverage extends MacroTransform with IdentityDenotTransformer:
           val statementId = recordStatement(parent, pos, false)
           insertInvokeCall(body, pos, statementId)
 
-    /** Returns the tree, prepended by a call to Invoker.invoker */
+    /** Returns the tree, prepended by a call to Invoker.invoked */
     private def insertInvokeCall(tree: Tree, pos: SourcePosition, statementId: Int)(using Context): Tree =
       val callSpan = syntheticSpan(pos)
       Block(invokeCall(statementId, callSpan) :: Nil, tree).withSpan(callSpan.union(tree.span))
 
-    /** Generates Invoked.invoked(id, DIR) */
+    /** Generates Invoker.invoked(id, DIR) */
     private def invokeCall(id: Int, span: Span)(using Context): Tree =
       val outputPath = ctx.settings.coverageOutputDir.value
       ref(defn.InvokedMethodRef).withSpan(span)
@@ -353,3 +370,15 @@ class InstrumentCoverage extends MacroTransform with IdentityDenotTransformer:
 object InstrumentCoverage:
   val name: String = "instrumentCoverage"
   val description: String = "instrument code for coverage checking"
+
+  /** Extractor object for trees produced by `insertInvokeCall`. */
+  object InstrumentedBlock:
+    private def isInvokedCall(app: Apply)(using Context): Boolean =
+      app.span.isSynthetic && app.symbol == defn.InvokedMethodRef.symbol
+
+    def unapply(t: Tree)(using Context): Option[(Apply, Tree)] =
+      t match
+        case Block((app: Apply) :: Nil, expr) if isInvokedCall(app) =>
+          Some((app, expr))
+        case _ =>
+          None

--- a/tests/coverage/pos/PolymorphicExtensions.scala
+++ b/tests/coverage/pos/PolymorphicExtensions.scala
@@ -1,0 +1,11 @@
+package covtest
+
+object PolyExt:
+  extension (s: String)
+    def foo[A](x: A): A = x
+
+  extension [A](i: Int)
+    def get(x: A): A = x
+
+  "str".foo(0) // ({foo("str")}[type])(0) i.e. Apply(TypeApply( Apply(foo, "str"), type ), List(0))
+  123.get(0) // {(get[type])(123)}(0) i.e. Apply(Apply(TypeApply(...), List(123)), List(0))

--- a/tests/coverage/pos/PolymorphicExtensions.scoverage.check
+++ b/tests/coverage/pos/PolymorphicExtensions.scoverage.check
@@ -1,0 +1,105 @@
+# Coverage data, format version: 3.0
+# Statement data:
+# - id
+# - source path
+# - package name
+# - class name
+# - class type (Class, Object or Trait)
+# - full class name
+# - method name
+# - start offset
+# - end offset
+# - line number
+# - symbol name
+# - tree name
+# - is branch
+# - invocations count
+# - is ignored
+# - description (can be multi-line)
+# '' sign
+# ------------------------------------------
+0
+PolymorphicExtensions.scala
+covtest
+PolyExt$
+Object
+covtest.PolyExt$
+foo
+61
+68
+4
+foo
+DefDef
+false
+0
+false
+def foo
+
+1
+PolymorphicExtensions.scala
+covtest
+PolyExt$
+Object
+covtest.PolyExt$
+get
+114
+121
+7
+get
+DefDef
+false
+0
+false
+def get
+
+2
+PolymorphicExtensions.scala
+covtest
+PolyExt$
+Object
+covtest.PolyExt$
+<init>
+138
+147
+9
+foo
+Apply
+false
+0
+false
+"str".foo
+
+3
+PolymorphicExtensions.scala
+covtest
+PolyExt$
+Object
+covtest.PolyExt$
+<init>
+138
+150
+9
+<none>
+Apply
+false
+0
+false
+"str".foo(0)
+
+4
+PolymorphicExtensions.scala
+covtest
+PolyExt$
+Object
+covtest.PolyExt$
+<init>
+238
+248
+10
+get
+Apply
+false
+0
+false
+123.get(0)
+

--- a/tests/coverage/pos/PolymorphicMethods.scala
+++ b/tests/coverage/pos/PolymorphicMethods.scala
@@ -1,0 +1,10 @@
+package covtest
+
+object PolyMeth:
+  def f[A](x: A): A = x
+  this.f(0) // (this.f[type])(0) i.e. Apply(TypeApply(Select(this,f), type), List(0))
+
+  C[String]().f("str", 0)
+
+class C[T1]:
+  def f[T2](p1: T1, p2: T2): Unit = ()

--- a/tests/coverage/pos/PolymorphicMethods.scoverage.check
+++ b/tests/coverage/pos/PolymorphicMethods.scoverage.check
@@ -1,0 +1,105 @@
+# Coverage data, format version: 3.0
+# Statement data:
+# - id
+# - source path
+# - package name
+# - class name
+# - class type (Class, Object or Trait)
+# - full class name
+# - method name
+# - start offset
+# - end offset
+# - line number
+# - symbol name
+# - tree name
+# - is branch
+# - invocations count
+# - is ignored
+# - description (can be multi-line)
+# '' sign
+# ------------------------------------------
+0
+PolymorphicMethods.scala
+covtest
+PolyMeth$
+Object
+covtest.PolyMeth$
+f
+36
+41
+3
+f
+DefDef
+false
+0
+false
+def f
+
+1
+PolymorphicMethods.scala
+covtest
+PolyMeth$
+Object
+covtest.PolyMeth$
+<init>
+60
+69
+4
+f
+Apply
+false
+0
+false
+this.f(0)
+
+2
+PolymorphicMethods.scala
+covtest
+PolyMeth$
+Object
+covtest.PolyMeth$
+<init>
+147
+158
+6
+<init>
+Apply
+false
+0
+false
+C[String]()
+
+3
+PolymorphicMethods.scala
+covtest
+PolyMeth$
+Object
+covtest.PolyMeth$
+<init>
+147
+170
+6
+f
+Apply
+false
+0
+false
+C[String]().f("str", 0)
+
+4
+PolymorphicMethods.scala
+covtest
+C
+Class
+covtest.C
+f
+187
+192
+9
+f
+DefDef
+false
+0
+false
+def f
+

